### ISSUE TITLE
Proposal: Add Ability to Restore Window Size and/or Position

### ISSUE THREAD

**Issue by robloo** [microsoft/microsoft-ui-xaml#2680]
# Proposal: Add Ability to Restore Window Size and/or Position

## Summary

Add the ability for an app to restore it's previous window size and position. This ideally should be provided out-of-the-box in WinUI 3.0. It also should be possible to customize by sitting a Window Size and Left/Top properties.

Implementation could be similar to this:
https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.restorebounds?view=netcore-3.1

## Rationale

Some apps are commonly used with larger or smaller window sizes than the default. There should be the ability to restore window size and position to the previous user-defined size when the app was last closed.

This will require the concept of a Window coming in WinUI 3.0.

## Scope

| Capability | Priority |
| :---------- | :------- |
| Add ability to restore window size and/or position with an app setting | Must |
| Add ability to set a custom window size and/or position during initialization | Should |

## Important Notes

* API to be discussed

**Comment by robloo** [comment]
Related to #2564

**Comment by StephenLPeters** [comment]
@marb2000 FYI

**Comment by marb2000** [comment]
Thanks! APIs to set position/move/size and restore are in our radar. No clear whether we can accomplish this for 2020. Meanwhile, Desktop WinUI 3 apps can use Win32 APIs like the ones used in the [WinUI-3-Demos](https://github.com/microsoft/WinUI-3-Demos/tree/master/DemoBuildCs)

**Comment by robloo** [comment]
Understood, great to hear its on the radar! I know a lot of focus is on WinUI/Win32; however, I work with UWP at the moment and would ideally have this functionally, at least in part, supported there as well. Something as simple as an option to automatically restore the previous window size by the OS would be nice (This wouldn't require any Window class or API additions other than 1 property on the Application class).

**Comment by Felix-Dev** [comment]
I would expect, just as with UWP apps today, that WinUI 3.0 UWP apps will automatically restore the size and position the app window last had when it was closed.

**Comment by robloo** [comment]
@Felix-Dev, does restoring app size work for you in Visual Studio UWP/C# apps during development? It doesn't on my end. You're right that UWP apps installed on the system seem to restore themselves just fine.

**Comment by Felix-Dev** [comment]
@robloo Yes, under the condition that I close the app normally (via corner-right close button, taskbar,...) and *not* by pressing the "stop debugging" button in VS. In the latter case, the last window size and position is not restored on next debug process.

**Comment by robloo** [comment]
@Felix-Dev, Interesting, it hasn't worked for me as long as I can recall -- even with a normal app shutdown. It might have something to do with the deferral I use to save state first. I'll have to play around with that some more. Thanks for mentioning it works on your end I never would have known otherwise.

**Comment by koenvd** [comment]
Is this still something on the radar? Or is there some sample code available on how to work around is? 
It's not really expected for a desktop app to not remember the last window position and size.

Thanks for feedback!

**Comment by castorix** [comment]
> Implementation could be similar to this:
https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.restorebounds?view=netcore-3.1

RestoreBounds mainly calls [GetWindowPlacement ](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getwindowplacement)

**Comment by rcohn** [comment]
Hi @castorix -

Thanks much for the pointer.  The code on saving and restoring to the user space will be very helpful.  However, the parts having to do with RestoreBounds are not supported for the WinUI3 app I'm building - I believe this would work in a UWP app.  I'll probably either do a pInvoke to GetWindowPlacement or something similar to get the actual Window location, bounds.

Thanks again.

**Comment by applefanbois** [comment]
If you save and restore the size and position yourself, it will cause problems if the screen no longer exist.

**Comment by applefanbois** [comment]
Of course not, and they never will fix it.

Windows is the only OS in the world that does offer a screen recorder with sound and mouse clicks.
Windows is the only desktop os that does not support tabs in the file explorer.
Windows is an abandoned platform for legacy software and nothing else.
Microsoft Store is the only store in the world that does not install app dependencies (does not install required dll from MS)???

**Comment by akshat157** [comment]
Hello! Are there any updates on this issue?
I noticed that WinUI3 has methods to `Move`, `Resize` and `MoveAndResize` an `AppWindow` which I used to implement a Window size and position restore (using app's local settings). However, it appears that this method does not work well on a multi-monitor setup and has issues with the display scaling.

- The `MoveAndResize()` (or perhaps the `Resize()` ) method appears to use the display scaling factor when it is used which leads to my manual `MoveAndResize()` increase or decrease the App's window size on each launch. The problem occurs on any scaling factor other than `100%`. Which makes me think that it is probably not possible to use these methods for this purpose.

**Comment by 27k1** [comment]
Hello! Are there any updates on this issue? I and [akshat157](https://github.com/akshat157) amongst many others require a fix for this please.

**Comment by tingjiangcao** [comment]
I think this is an urgent issue that needs to be addressed.
It's a bug, not an "optimization issue."

But anyone who has used desktop apps, including the old wpf or macos next door, has done this. Because this is a must for users, not an option.

Please Microsoft prioritize this type of issue with the ultimate customer experience in mind. This is seriously interfering with smooth development for developers.

**Comment by tingjiangcao** [comment]
I think this is a problem that needs to be addressed urgently.
It's a "bug", not an "optimization issue"

Anyone who has ever used a desktop application (including older versions of wpf or macos) knows that this is a must-have, most basic feature for users.

From the release of winui3 until now, no attention has been paid to this issue and resolved.

Please Microsoft prioritize this type of issue with the ultimate customer experience in mind. This has seriously interfered with the smooth development of developers.

**Comment by applefanbois** [comment]
It is almost impossible to make it work with multi screen with different scaling.
Because windows are always created on screen zero then moved to another window, it is impossible to know the scaling of the screen where it is going to be.

When my clients complains, I tell them : get the Apple version.

At this point I am so tired of fighting an abandoned platform that I am focusing on the Android version, because that version will work natively on Windows 11 and the android version won't have any bugs and will get all the features that the apple version has.

In the presence of a Webview2, it is impossible to intercept keyboard events (because of a MS bug). So all the keyboard shortcuts do not work on the windows version. Microsoft knows about that and has not resolved the issue in 3 years.

**Comment by AkazaRenn** [comment]
> > I would really appreciate having such basic features to be implemented before adding all those AI stuff.
> 
> In this thread : https://github.com/microsoft/WindowsAppSDK/issues/5896
> @ghost1372 said that API was working.

That's great to hear! I read the documentary and it does look promising. Can't wait to see it getting released.
